src/compiler/checker.ts

```diff
@@ -25110,6 +25233,48 @@ namespace ts {
             return literalType;
         }
 
+        /**
+         * Replaces all references to structurally equivalent object and array literal types in the given list
+         * with references to a single one of those types. The process is applied recursively to properties of
+         * object literals and elements of array literals.
+         */
+        function deduplicateObjectOrArrayLiteralTypes(types: Type[]) {

```

Comment:
We had discussed issues around circular types in object literal methods/getters in a past design meeting - do those concerns not apply here?

Reply:
They do not. See my reply above.